sr_vision_mocks/src/sr_vision_mocks/mock_recognize_gazebo_objects.py

```diff
@@ -0,0 +1,71 @@
+#!/usr/bin/env python
+
+#   This node listens to Gazebo's published model states, and publishes a TF
+#   for each object, as well as a RecognizedObjectArray. It is intended to
+#   mock a perfect vision solution, in order to test grasping etc.
+
+import rospy
+import threading
+import tf2_ros
+from gazebo_msgs.msg import ModelStates
+from geometry_msgs.msg import TransformStamped
+from object_recognition_msgs.msg import RecognizedObjectArray, RecognizedObject
+
+
+class MockRecognizeGazeboObjects(object):
+    def __init__(self):
+        self.recognized_object_array = RecognizedObjectArray()
+        self.tfs = []
+        self.lock = threading.Lock()
+        self.br = tf2_ros.TransformBroadcaster()
+        self.r = rospy.Rate(10)
+        rospy.Subscriber("/gazebo/model_states", ModelStates, self.on_model_states)
+        self.publisher = rospy.Publisher('/recognized_objects', RecognizedObjectArray,
+                                         queue_size=1)
+        while not rospy.is_shutdown():
+            self.r.sleep()
+            self.publish()
+
+    def publish(self):
+        self.lock.acquire()
```

Comment:
this might be my limited knowledge of locks here but is locking actually necessary if you're not using threading?

Reply:
If I understand correctly, rospy callback threading is different to roscpp. In rospy subscriber callbacks get their own thread, so it is possible that the on_model_states() callback would happen and attempt to write to self.tfs while publish() is attempting to read it.